File: model/src/main/java/io/github/runtimemodels/chazm/relation/PossessesFactory.java
package io.github.runtimemodels.chazm.relation;

import io.github.runtimemodels.chazm.entity.Agent;
import io.github.runtimemodels.chazm.entity.Capability;

/**
 * The {@linkplain PossessesFactory} interface defines the API for constructing {@linkplain Possesses} relations.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
public interface PossessesFactory {

	/**
	 * Constructs a {@linkplain Possesses}.
	 *
	 * @param agent
	 *            the {@linkplain Agent} of the {@linkplain Possesses}.
	 * @param capability
	 *            the {@linkplain Capability} of the {@linkplain Possesses}.
	 * @param score
	 *            the <code>double</code> score of the {@linkplain Possesses}.
	 * @return a {@linkplain Possesses}.
	 */
	Possesses buildPossesses(Agent agent, Capability capability, double score);

}

File: chazm-model/src/main/kotlin/runtimemodels/chazm/model/relation/PossessesFactory.kt
package runtimemodels.chazm.model.relation

import runtimemodels.chazm.api.entity.Agent
import runtimemodels.chazm.api.entity.Capability
import runtimemodels.chazm.api.relation.Possesses

/**
 * The [PossessesFactory] interface defines the API for constructing [Possesses] relations.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
interface PossessesFactory {
    /**
     * Constructs a [Possesses].
     *
     * @param agent      the [Agent] of the [Possesses].
     * @param capability the [Capability] of the [Possesses].
     * @param score      the `double` score of the [Possesses].
     * @return a [Possesses].
     */
    fun build(agent: Agent, capability: Capability, score: Double): Possesses
}
